### ISSUE TITLE
`from __future__ import barry_as_FLUFL` doesn't work

### ISSUE THREAD

**Issue by JelleZijlstra** [python/cpython#125331]
# Bug report

### Bug description:

```
% ./python.exe -c 'from __future__ import barry_as_FLUFL; print(1 <> 2)'
  File "<string>", line 1
    from __future__ import barry_as_FLUFL; print(1 <> 2)
                                                   ^^
SyntaxError: invalid syntax
```

But with Barry as the FLUFL, `<>` is the correct way to test for inequality.

The future works only if you pass the right flag to `compile()` (as test_flufl.py does), but there's no way to do that in a file.

Obviously this future is a joke so this doesn't matter greatly, but I think it indicates that future imports that affect the parser aren't handled properly. That's something we should fix in case we get other futures in the ... future that also affect the parser. The joke future is also useful as a way to demonstrate how future statements work in general without having to rely on a specific future that will go away in a few releases.

### CPython versions tested on:

3.12, CPython main branch

### Linked PRs
* gh-125482
* gh-131062
* gh-131063

**Comment by nineteendo** [comment]
We would have to be careful to make sure that this easter egg can't be used in production...

**Comment by Wulian233** [comment]
This issue was discussed recently, and IDLE is not a bug
https://github.com/python/cpython/pull/124999#discussion_r1789081911

https://github.com/python/cpython/pull/124999#discussion_r1789087389

**Comment by ZeroIntensity** [comment]
Clearly this should be a blocker for all upcoming releases 😄

**Comment by pablogsal** [comment]
I'm investigating and working on a possible fix. I am not fully convinced that is a but but I agree it looks like one.

**Comment by pablogsal** [comment]
> Obviously this future is a joke so this doesn't matter greatly, but I think it indicates that future imports that affect the parser aren't handled properly.

Not really: the parser receives all its configuration when executed, including the flags. The parser cannot change the configuration mid execution (among other things because it doesn't "know" what an import is supposed to do).

> That's something we should fix in case we get other futures in the ... future that also affect the parser. 

The way most futures work is by changing the compiler, not the parser. This is the exception and in reality is not even an exception because the way this was added was to only work in interactive mode. Indeed if you try this in a file:

```
from __future__ import barry_as_FLUFL
1 <> 2
```

it won't work:

```
 File "/home/pablogsal/github/python/main/lol.py", line 2
    1 <> 2
      ^^
SyntaxError: invalid syntax
```

Is only prepared to work on sequential-execution where the previous statement modifies the flags. That's why it works in the REPL:

```
>>> from __future__ import barry_as_FLUFL
>>>
>>> 1 <> 2
True
```

this is because when executing the first statement, the second call to parse changed flags. But this only happened after it was **executed** (not parsed).

**Comment by pablogsal** [comment]
The toll to fix this is not trivial: every call to parse/compile would need to be actually 2 calls to parse/compile OR we need to teach the parser to change flags mid run (which is not trivial because future imports have certain restrictions):

```
1 != 2
from __future__ import barry_as_FLUFL
```

gives:

```
  File "/home/pablogsal/github/python/main/lol.py", line 2
    from __future__ import barry_as_FLUFL
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: from __future__ imports must occur at the beginning of the file
```

and these restrictions are not checked by the parser but the compiler (in the symbol table):

https://github.com/python/cpython/blob/843d28f59d2616d052d9d45f31823976da07f0f3/Python/symtable.c#L1777

**Comment by pablogsal** [comment]
I think that given that this will introduce some slowness to fix or maintenance burden in the parser and **this was the way this worked since it was introduced** I suggest to close as not a bug.

**Comment by pablogsal** [comment]
Ok I have a "future-hack" PEG edition:

https://github.com/python/cpython/pull/125482

We could go with this now if @lysnikolaou doesn't find it too gross.

**Comment by thatbirdguythatuknownot** [comment]
I wanted to make this issue along with the other `barry_as_FLUFL` issue I found a few months ago but I thought it would get shut down immediately :p

**Comment by wimglenn** [comment]
Coincidentally, I asked similar on stack overflow back in Sept- https://stackoverflow.com/q/79004026/674039 

Nobody was able to answer it there (yet), but Pablo's comments here effectively answer the question.
